sql/src/main/java/org/apache/druid/sql/calcite/schema/DruidSchema.java

```diff
@@ -431,19 +431,29 @@ void removeSegment(final DataSegment segment)
     }
   }
 
-  private void removeServerSegment(final DruidServerMetadata server, final DataSegment segment)
+  @VisibleForTesting
+  void removeServerSegment(final DruidServerMetadata server, final DataSegment segment)
   {
     synchronized (lock) {
       log.debug("Segment[%s] is gone from server[%s]", segment.getId(), server.getName());
       final Map<SegmentId, AvailableSegmentMetadata> knownSegments = segmentMetadataInfo.get(segment.getDataSource());
       final AvailableSegmentMetadata segmentMetadata = knownSegments.get(segment.getId());
-      final Set<String> segmentServers = segmentMetadata.getReplicas();
-      final ImmutableSet<String> servers = FluentIterable.from(segmentServers)
-                                                         .filter(Predicates.not(Predicates.equalTo(server.getName())))
-                                                         .toSet();
+      final Set<DruidServerMetadata> segmentServers = segmentMetadata.getReplicas();
+      final ImmutableSet<DruidServerMetadata> servers = FluentIterable
+          .from(segmentServers)
+          .filter(Predicates.not(Predicates.equalTo(server)))
+          .toSet();
+      final Optional<DruidServerMetadata> realtimeServer = servers
+          .stream()
+          .filter(metadata -> metadata.getType().equals(ServerType.REALTIME))
+          .findAny();
+
+      // if there is no realtime server in the replicas, isRealtime flag should be unset
+      long isRealtime = realtimeServer.isPresent() ? 1 : 0;
```

Comment:
[Apologies for missing this original discussion](https://github.com/apache/incubator-druid/pull/6094#discussion_r210661609), but is this the most useful thing for the `is_realtime` flag to mean, where it reports true if _any_ of the servers serving a segment is a realtime task?

It might be more useful for `is_realtime` to report true if it is _only_ available on realtime tasks, and false if any server serving the segment is a historical, since that is how the broker will behave at query time (#4764). I think this would probably involve doing this as part of `addSegment` instead of `removeSegment`.

Reply:
You have a point, i think i sort of agree, it'll make `is_realtime` more intuitive for users.  The docs for `is_realtime` [here](https://druid.apache.org/docs/latest/querying/sql) say this flag is `1 if this segment is being served on any type of realtime tasks`, that'll need to change. Yeah, with the changed semantics of `is_realtime`, the flag will be unset when `addSegment` is called a second time via a historical. May be it should be updated in `removeSegment` as well, but the check will be if there is any historical server in the replicas, unset this flag.

Reply:
This change will make it `incompatible` with previous behavior, which I hope is okay, unless someone is relying on that behavior of `is_realtime`.

Reply:
> It might be more useful for is_realtime to report true if it is only available on realtime tasks, and false if any server serving the segment is a historical

That sounds good to me

Reply:
I think the idea I had in the previous conversation was motivated by `is_realtime` being false meaning that the segment has completely flushed out of the realtime system. I thought it'd be useful for finding segments where handoff is lingering for some reason.

If we change it to what you are proposing, @clintropolis, then it'd be somewhat harder to find segments where handoff is lingering (but still possible by joining `segments` and `servers`). But it'd be easier to find segments where handoff has progressed far enough that a historical has loaded the segment.

Six of one, half dozen of the other.

I am okay with either approach, so long as it's clearly documented what the behavior should be. I don't think the fact that the change would be theoretically incompatible is a big deal, since the column is buggy enough that it doesn't work anyway right now.